File: leetcode/recursion/fibonacci/fib.java
import java.util.HashMap;

class Solution {
    HashMap<Integer, Integer> cache = new HashMap<Integer, Integer>();
    
    public int fib(int n) {
        if (cache.containsKey(n)) {
            return cache.get(n);
        }
        int ans;
        
        if (n <= 1) {
            ans = n;
        } else {
            ans = fib(n-1) + fib(n-2);
        }
        cache.put(n, ans);
        return ans;
        
    }
}

File: leetcode/recursion/fibonacci/fib.py
def fib(n: int) -> int:
    cache = {}
    
    def cache_fib(N):
        print(cache)
        if N in cache:
            cache[N]
        
        if N <= 1:
            ans = N
        else:
            f1 = cache_fib(N-1)
            f2 = cache_fib(N-2)
            print("f1: ", f1, "f2: ", f2)
            ans = f1 + f2
        cache[N] = ans
        return ans
    return cache_fib(n)